### ISSUE TITLE
MSL: Inline uniform blocks in Metal argument buffer containing structs need padding if std140 alignment

### ISSUE THREAD

**Issue by billhollings** [KhronosGroup/SPIRV-Cross#1572]
When using the `ComplierMSL::add_inline_uniform_block()` function to embed block data directly into a Metal argument buffer, if `std140` alignment is enabled, structs and array elements are required to be aligned (padded) to 16 bytes. However, Metal's default struct alignment is simply that of its member with the largest alignment.

For example, in the following generated MSL, Metal will assume the size and alignment of the first two members of `spvDescriptorSetBuffer0` is 4 bytes, but GLSL, SPIR-V, and Vulkan assume they are 16 bytes:

```
struct _10
{
    int _m0;
};

struct _22
{
    int _m0;
};

struct _30
{
    int _m0;
};

struct spvDescriptorSetBuffer0
{
    _10 m_12 [[id(0)]];
    _22 m_24 [[id(1)]];
    device _30* m_32 [[id(2)]];
};
```

SPIR-V, MSL & GLSL conversion output for the source of this example is here: [MSLArgBuff-InlineUniformBlock.txt](https://github.com/KhronosGroup/SPIRV-Cross/files/5726526/MSLArgBuff-InlineUniformBlock.txt)

If `std140` alignment is active, these structs need to be aligned or padded to 16 bytes in the generated MSL. This is not required if using `std430` alignment. 

I expect this behaviour should also be optional via `CompilerMSL::Options` to allow it to be turned off if the Vulkan `uniformBufferStandardLayout` feature has been enabled.

I wonder if this padded alignment might be done using the `alignas` specifier on the members of `spvDescriptorSetBufferN` that are structs, using the minimum of the largest struct element or 16 when `std140` is active.

**Comment by cdavis5e** [comment]
> I wonder if this padded alignment might be done using the `alignas` specifier on the members of `spvDescriptorSetBufferN` that are structs, using the minimum of the largest struct element or 16 when `std140` is active.

This will align the struct itself to that. We need to align the *members* of that struct that are themselves structs--or arrays.

Unfortunately, it won't work for arrays, because for arrays we need to align the element itself, not the array as a whole. There is no way to make `alignas` apply to an array element. I've filed FB8847491 asking for this.

**Comment by billhollings** [comment]
>This will align the struct itself to that. We need to align the members of that struct that are themselves structs--or arrays.

Which struct are you saying will be aligned? I was suggesting to add the `alignas` to each member of `spvDescriptorSetBufferN` that are themselves structs, expecting that it would align only that member (ie- pad between members of `spvDescriptorSetBufferN`). 

Are you indicating the entire `spvDescriptorSetBufferN` would be what is aligned? Therefore we'd need to add the `alignas` to say the `struct _10` definition above?

And yes...arrays are even more problematic and harder to deal with. I wonder if putting the `alignas` on the struct definition (ie- `struct _10` above) would solve that?

**Comment by billhollings** [comment]
instead of using `alignas`, we could also just add appropriate padding members between the actual members of `spvDescriptorSetBufferN`, as we sometimes do elsewhere in MSL generation.

**Comment by cdavis5e** [comment]
> Are you indicating the entire `spvDescriptorSetBufferN` is what would be aligned?

No. I'm saying that, when we have an inline uniform block, the struct for the block is embedded in the argument buffer. *That's* the struct that will get aligned, and *that's* the struct whose composite members need explicit alignment.

> Therefore we'd need to add the `alignas` to say the `struct _10` definition above?

No. We'd need to add it to the *members* of `struct _10`. And `struct _22`.
 
> And yes...arrays are even more problematic and harder to deal with. I wonder if putting the `alignas` on the struct definition (ie- `struct _10` above) would solve that?

If the IUB itself is arrayed, that might work. If the IUB has an array that is a member, that's a problem--especially if that member's elements are of a scalar or vector type.

**Comment by billhollings** [comment]
>We'd need to add it to the members of `struct _10`.

Okay. Gotcha. In my example, I was dealing only with the alignment of the elements of the Metal argument buffer itself, not with the members of those elements. Fair enough...and good catch.

So we'll need to do both. Because in the example above, we don't need to align the `int _m0`, but we do need to align the members of `spvDescriptorSetBuffer0`.

**Comment by HansKristian-Work** [comment]
Is it even required to support more than one inline uniform buffer per descriptor set? If only one inline UBO is used, there is no alignment issue I think?

**Comment by billhollings** [comment]
> Is it even required to support more than one inline uniform buffer per descriptor set? If only one inline UBO is used, there is no alignment issue I think?

The SPIR-V that I included a link for above includes a descriptor set that contains two Block bindings. These become inline structs in an MSL argument buffer.

I've worked around this issue for now by using temporary MTLBuffers for blocks. Once things are farther along, we can revisit this further.
